### DISCUSSION TITLE
How to programmatically monitor a sensor for failures

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#13330] [Q&A]
Who watches the watchmen?

If your sensor has complex logic, and you want to catch failures and do something with them, here's how. This is for failures in the sensor itself, not failures in the runs launched by the sensor. Note that these APIs are currently not public and may change in the future (though likely slowly, because storage is difficult to change).

### Step 1: find your sensor's origin_id and selector_id

To do this, you'll need to find the `jobs` table in your instance database, then query it to find the `origin_id` and `selector_id` of your sensor.

```sql
select job_origin_id, selector_id
from jobs
where instr(job_body, "my_sensor") != 0
```

Replace `"my_sensor"` with the name of your sensor. The above uses SQLite SQL. If using Postgres, you'll need to replace the last line with:

```sql
where job_body like "%my_sensor%"
```

### Step 2: query the instance DB for failing ticks

```python
from typing import Sequence
from dagster import DagsterInstance
from dagster._core.scheduler.instigation import TickStatus, InstigatorTick


def print_sensor_failures(instance: DagsterInstance, cursor: float):
    failing_ticks: Sequence[InstigatorTick] = instance.get_ticks(
        origin_id="3c1ffb29b2bb7b3b98baa89e58755e5bed0691f5", # use the origin_id from above
        selector_id="6bb42a51de68252fbb61ee5e305ab400f1477b43", # use the selector_id from above
        after=cursor,
        statuses=[TickStatus.FAILURE],
        limit=5,
    )

    for tick in failing_ticks:
        print(tick.error.to_string())


if __name__ == "__main__":
    instance = DagsterInstance.get()
    print_sensor_failures(instance, 0.0)
```

The above code could also go inside another Dagster sensor.

**Comment by abkfenris** [comment]
You could also have the sensor ping a cron monitoring service like Healthchecks.io, Croniter, or Dead Man's Snitch after yielding runs.
